src/statistics/ranks.rs

```diff
@@ -10,146 +7,99 @@ pub trait Ranks<T> {
 impl<T> Ranks<f64> for T
 where
     T: IntoIterator,
-    T::Item: Borrow<f64>,
+    T::Item: PartialOrd + Copy + Default,
 {
     fn ranks(self) -> (Vec<f64>, usize) {
-        let mut observations: Vec<_> = self.into_iter().map(|x| *x.borrow()).enumerate().collect();
-        observations
-            .sort_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
-
-        let (sorted_indices, sorted_values): (Vec<_>, Vec<_>) = observations.into_iter().unzip();
-        let groups = DedupWithCount::new(sorted_values.iter());
+        let mut observations: Vec<(usize, T::Item)> = self.into_iter().enumerate().collect();
+        observations.sort_unstable_by(|(_, a), (_, b)| {
+            a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal)
+        });
 
-        let mut resolved_ties = ResolveTies::new(groups);
-        let mut ranks = vec![0.0; sorted_values.len()];
+        let mut resolved_ties = ResolveTies::from(observations.iter().map(|(_, value)| *value));
+        let mut ranks = vec![0.0; observations.len()];
 
-        for (rank, old_index) in (&mut resolved_ties).zip(sorted_indices) {
+        for ((rank, _), old_index) in
+            (&mut resolved_ties).zip(observations.iter().map(|(index, _)| *index))
+        {
             ranks[old_index] = rank;
         }
 
         (ranks, resolved_ties.tie_correction())
     }
 }
 
-struct ResolveTies<I, T>
+pub(crate) struct ResolveTies<I, F>
 where
-    I: Iterator<Item = (usize, T)>,
+    I: Iterator,
 {
     iter: I,
     index: usize,
-    left: usize,
-    resolved: Option<f64>,
+    resolved: f64,
     tie_correction: usize,
+    current_normalized_item: Option<I::Item>,
+    normalize: F,
 }
 
-impl<I, T> ResolveTies<I, T>
+impl<I, F> ResolveTies<I, F>
 where
-    I: Iterator<Item = (usize, T)>,
+    I: Iterator,
 {
-    fn new(iter: I) -> Self {
+    pub(crate) fn new(iter: I, normalize: F) -> Self {
         ResolveTies {
             iter,
             index: 0,
-            left: 0,
-            resolved: None,
+            resolved: 0.0,
+            current_normalized_item: None,
             tie_correction: 0,
+            normalize,
```

Comment:
There is no public code path that lerads to normalized to be set to anything but the identity function, would it make sense to skip it and not pay the penalty for it?

Reply:
Normalized is set to 'abs' in the Wilcoxon test [here](https://github.com/larsgw/stattest/blob/8c9c533bc7008d4cd811ea78579a4accb11f9a73/src/test/wilcoxon_w.rs#L316) to avoid rendering the absolute valued vector, which otherwise would force us to store separately the sign and keep it aligned with the absolute-values variant when we compute the rank. In the Mann-Whitney test, instead, we only need to use the identity. Therefore, to avoid code redundancy, I provided that as a function pointer. I believe it may be still inlined by the compiler, though I would need to verify it.

Reply:
I'm not sure it'd completely compile it out OTOH rust is amazing at not doing work :joy: so it could just be the case.

But especially in the `iter.clone().map(&self.normalize).take` case I'd be surprised (but I like being surprised :D)

Reply:
I will benchmark it and see whether anything changes and let you know

Reply:
Harcoding the normalization function, i.e. writing directly the abs in place of the `self.normalize` does not change the performance according to the Criterion benchmark line up.

Reply:
I am more thinking about that we do not need normalize at all outside of the test from what I can tell.

Reply:
Sorry to be dense but I am not sure I understood that - could you kindly expand upon it?

Reply:
I think I totally got fooled by naming `Wilcoxon test` is the `Wilcoxon test` method not the test for the `Wilcoxon` method, disregard everything I stayed 🤣